### ISSUE TITLE
[bug] May be a bug of gradients calculation.

### ISSUE THREAD

**Issue by nianyou** [tensorflow/tensorflow#20613]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**:NO
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:Linux Ubuntu 16.04
- **TensorFlow installed from (source or binary)**:binary
- **TensorFlow version (use command below)**:1.9
- **Python version**: 2.7
- **Bazel version (if compiling from source)**:
- **GCC/Compiler version (if compiling from source)**:
- **CUDA/cuDNN version**:
- **GPU model and memory**:
- **Exact command to reproduce**:

I write a test case to test operation ops::Square and it's gradient function SquareGrad, here is the code:

    TEST(MathTest, SingleComplex64Function) {
     //
     //data prepare
     Scope root = Scope::NewRootScope();
     DataType x_type = DataTypeToEnum<complex64>::v();
     TensorShape shape({1});

     Tensor x_data(x_type, shape);
     auto x_data_flat = x_data.flat<complex64>();
     x_data_flat(0) = {1,2};
     
     //
     //graph construct: y = x*x, x and y are complex64
     auto x = ops::Placeholder(root, x_type, Placeholder::Shape(shape));
     auto y = ops::Square(root, x);

     ClientSession session(root);
     std::vector<Tensor> outputs;

     //
     //calculate y
     Status s = session.Run({{x, x_data}} , {y}, &outputs);
     ASSERT_TRUE(s.ok());
     ASSERT_TRUE(outputs.size() == 1);
     ASSERT_TRUE(outputs[0].dtype() == DT_COMPLEX64);
     ASSERT_TRUE(outputs[0].flat<complex64>()(0) == complex64(-3,4));

     std::vector<Output> grads;
     s = AddSymbolicGradients(root, {y}, {x}, &grads);
     ASSERT_TRUE(s.ok());
     
     //
     //calculate gradients of y w.r.t x
     std::vector<Tensor> grad_outputs;
     s = session.Run({{x, x_data}}, grads, &grad_outputs);
     ASSERT_TRUE(s.ok());
     ASSERT_TRUE(grad_outputs.size() == 1);
     ASSERT_TRUE(grad_outputs[0].dtype() == DT_COMPLEX64);
     ASSERT_TRUE(grad_outputs[0].flat<complex64>()(0) == complex64(2,4));
    }

when i run it, i get this error:

    [ RUN      ] MathTest.SingleComplex64Function
    tensorflow/cc/gradients/math_grad_test.cc:84: Failure
    Value of: grad_outputs[0].flat<complex64>()(0) == complex64(2,4)
      Actual: false
      Expected: true
    [  FAILED  ] MathTest.SingleComplex64Function (39 ms)

I find the calculated value of grad_outputs[0] is (2, -4) , is this correct?

According to my knowledge, i have defined a function R -> R : y = x^2 (x and y are complex64),  the first derived function of it should be : dy/dx = 2x, and according to the definition of gradient: grad y(x) = [dy/dx], so if z=1+2i, y should be (1+2i)(1+2i)=1+4i+4i*i =1+4i-4=-3+4i,  and dy/dx = 2(1+2i)=2+4i, so theoretically grad y should be [(2,4)].

**Comment by nianyou** [comment]
I have found a issue talking about the similar problem: https://github.com/tensorflow/tensorflow/issues/3850
but still i don't understand girving's explanation.

Suppose x = x_real + i * x_imag, y = y_real + i * y_imag,  i have also found out what AddSymbolicGradients really try to calculate is something like z = dy_real/dx_real + i * dy_real/dx_imag. 

Is this the definition of gradient for complex-valued function in tensorflow ?

**Comment by jony0917** [comment]
This is not a bug, and yes , what is really output is dy_real/dx_real + i * dy_real/dx_imag for complex-valued functions y=f(x).

I think the key to your question is the definition of gradient. let me explain it.

* for real-valued R -> R funciton f:  y = f(x) , the gradient of f w.r.t x is grad f(x) = df/dx,  

Suppose what we aim to minimize a graph's output L, and L = z, z = z(y), y = y(x), we know that dz/dx = dz/dy * dy/dx, 
so we can calculate gradient w.r.t any variant in this graph like this :

```
grad L(z) = dL/dz = 1
grad L(y) = dL/dy = dL/dz * dz/dy = grad L(z) * dz/dy
grad L(x) = dL/dx = dL/dy * dy/dx = grad L(y) * dy/dx
```

this is called the reversed calculation of gradients.


But for complexed-valued function C -> C: z = f(w),  if we define gradient f(w) = dz/dw, the chain rule can not apply directly.

so, we change the definition:

* for complex-valued C -> C function f: z = f(w), we define two gradient: the real gradient of f w.r.t w is grad_real f(w) = dz_real/dw_real + i * dz_real/dw_imag, 
the imaginary gradient of f w.r.t w is grad_imag f(w) = dz_imag/dw_real + i * dz_imag/dw_imag.


Now, let's see how to calcualte the real gradient and the imag gradient.

Similarly suppose we get a graph: L = z, z = z(y), y = y(x),  L,z,y,x are all complex variants. By the definition, we do it like this:

```
grad_real L(z) = (dz_real/dz_real + i * dz_real/dz_imag) = (1, 0)
grad_imag L(z) = (dz_imag/dz_real + i * dz_imag/dz_real) = (0, i)

grad_real L(y) = (dz_real/dy_real + i * dz_real/dy_imag) 
= (1, 0) * (dz_real/dy_real + i * dz_real/dy_imag)

And we know that for a analytic fucntion z = z(y) , wo get Cauchy–Riemann equations:
dz_real/dy_real = dz_imag/dy_imag
dz_real/dy_imag = - dz_imag/dz_real

and we know that :
dz/dy = dz_real/dy_real + i * dz_imag/dz_real
      = dz_real/dy_real - i * dz_real/dy_imag
      = dz_imag/dy_imag + i * dz_imag/dy_real
      = dz_imag/dy_imag - i * dz_real/dy_imag

so, 
grad_real L(y) = (1, 0) * (dz_real/dy_real + i * dz_real/dy_imag) 
               = grad_real(z) * conjugate(dz_real/dy_real - i * dz_real/dy_imag)
               = grad_real(z) * conjugate(dz/dy)


grad_imag L(y) = (dz_imag/dy_real + i * dz_imag/dy_imag) 
               = (0, i) * (dz_imag/dy_imag - i * dz_imag/dy_real)
               = grad_imag(z) * conjugate(dz_imag/dy_imag + i * dz_imag/dy_real)
               = grad_imag(z) * conjugate(dz/dy)



grad_real L(x) = dz_real/dx_real + i * dz_real/dx_imag
= (dz_real/dy_real * dy_real/dx_real + dz_real/dy_imag * dy_imag/dx_real) + i * (dz_real/dy_real*dy_real/dx_imag + dz_real/dy_imag * dy_imag/dx_imag)
= dz_real/dy_real * (dy_real/dx_real + i * dy_real/dx_imag) +  dz_real/dy_imag * (dy_imag/dx_real + i * dy_imag/dx_imag)
= dz_real/dy_real * (dy_real/dx_real + i * dy_real/dx_imag) +  i * dz_real/dy_imag * (dy_imag/dx_imag - i * dy_imag/dx_real)
= dz_real/dy_real * conjugate(dy_real/dx_real - i * dy_real/dx_imag) + i * dz_real/dy_imag * conjugate(dy_imag/dx_imag + i * dy_imag/dx_real)
= dz_real/dy_real * conjugate(dy/dx) + i * dz_real/dy_imag * conjugate(dy/dx)
= (dz_real/dy_real + i * dz_real/dy_imag) *  conjugate(dy/dx)
= grad_real(y) * conjugate(dy/dx)



grad_real L(x) = dz_imag/dx_real + i * dz_imag/dx_imag
= (dz_imag/dy_real * dy_real/dx_real + dz_imag/dy_imag * dy_imag/dx_real) + i * (dz_imag/dy_real * dy_real/dx_imag + dz_imag/dy_imag * dy_imag/dx_imag)
= dz_imag/dy_real * (dy_real/dx_real + i * dy_real/dx_imag) + dz_imag/dy_imag * (dy_imag/dx_real + i * dy_imag/dx_imag)
= dz_imag/dy_real * (dy_real/dx_real + i * dy_real/dx_imag) + i * dz_imag/dy_imag * (dy_imag/dx_imag - dy_imag/dx_real)
= dz_imag/dy_real * conjugate(dy_real/dx_real - i * dy_real/dx_imag) + i * dz_imag/dy_imag * conjugate(dy_imag/dx_imag + dy_imag/dx_real)
= dz_imag/dy_real * conjugate(dy/dx) + i * dz_imag/dy_imag * conjugate(dy/dx)
= (dz_imag/dy_real + i * dz_imag/dy_imag) * conjugate(dy/dx)
= grad_imag(y) * conjugate(dy/dx)
```

Summing up, for the graph: L = z, z = z(y), y = y(x), for any real variant t, we define L gradient w.r.t t is Grad L(t) = grad L(t), for any 
complex variant t we define L gradient w.r.t t Grad L(t) = grad_real L(t); because for any real variant x, x = conjugate(x), now we can calculate Grad L w.r.t any variant by the same chain rules :

```
Grad L(z) = 1,
Grad L(y) = Grad L(z) * conjugate(dz/dy)
Grad L(x) = Grad L(y) * conjugate(dy/dx)
```

**Comment by tensorflowbutler** [comment]
Nagging Assignee @bignamehyp: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by bignamehyp** [comment]
Can we close this issue?

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this still an issue?

**Comment by tensorflowbutler** [comment]
It has been 29 days with no activity and the `awaiting response` label was assigned. Is this still an issue?

**Comment by tensorflowbutler** [comment]
It has been 44 days with no activity and the `awaiting response` label was assigned. Is this still an issue?

**Comment by ymodak** [comment]
Automatically closing due to lack of recent activity. Please update the issue when new information becomes available, and we will reopen the issue. Thanks!
